### PR TITLE
net-misc/i2pd: bugfix & housecleaning 

### PR DESCRIPTION
**Pull request by khumarahn**
This PR replaces https://github.com/gentoo/gentoo/pull/1222 after receiving a lot of feedback from @tomboy-64 and @blueness.

* 3e8340f: config files are now owned by `root:i2pd`, `i2pd` group has write access. I would be in favour of not allowing `i2pd` group write access, because then the daemon has it being run as  `i2pd:i2pd`. On the other hand, this way users with membership in `i2pd` can change config, which is good. Though, they still cannot restart the service so the changes take effect...
* 51b5ab8: we remove unused /usr/share/i2pd
* 1a7027b: restructure the install section of the ebuild, purely cosmetic
* 0337b1c: fix pid file location for logrotate (this is a bugfix)
* bab5896: make i2pd user and group static (not depending on environment variables). I tend to think that we need this, because systemd service uses static names anyway, so we did not guarantee that the environment variables `I2PD_USER` and `I2PD_GROUP` were correctly used for systemd users. Also, this way it is simpler.
* 2effb25: save virgin copies of config files in `/usr/share/doc/i2pd-2.6.0-r3/`, along with `README.md`, which was already placed there
* 55a3684: remove spaces in retry for start-stop-daemon, as in https://github.com/OpenRC/openrc/issues/76
* 6dda70b: increase max number of open files (`ulimit -n`) to 4096. The default value of 1024 is not enough for floodfill: daemon becomes unresponsive, and webconsole `http://localhost:7070` does not open.

### CODE REVIEW

File: net-misc/i2pd/i2pd-2.6.0-r3.ebuild [hunk 1]
```diff
@@ -72,6 +72,9 @@ src_install() {
 
 	# doc
 	dodoc README.md
+	dodoc docs/i2pd.conf
+	dodoc debian/tunnels.conf
+	dodoc debian/subscriptions.txt
```
Comment (tomboy-64) [inline net-misc/i2pd/i2pd-2.6.0-r3.ebuild]: > DOCS 	An array or space-delimited list of documentation files for the default src_install function to install using dodoc. If undefined, a reasonable default list is used. (Requires EAPI>=4.)

Just put those into the global array like this: ```DOCS=( README.md docs/i2pd.conf ... )```

File: net-misc/i2pd/i2pd-2.6.0-r3.ebuild [hunk 2]
```diff
@@ -72,12 +72,16 @@ src_install() {
...
-	fperms 600 "/etc/${PN}/${PN}.conf" \
-		"/etc/${PN}/subscriptions.txt" \
-		"/etc/${PN}/tunnels.conf"
+
+	# grant i2pd group write access to config files
+	fowners "root:${I2PD_GROUP}" \
+		/etc/i2pd/i2pd.conf \
+		/etc/i2pd/tunnels.conf \
+		/etc/i2pd/subscriptions.txt
+	fperms g+w \
```
Comment (tomboy-64) [inline net-misc/i2pd/i2pd-2.6.0-r3.ebuild]: Please use an explicit octal mask here, e.g. ```660```.

File: net-misc/i2pd/i2pd-2.6.0-r3.ebuild [hunk 3]
```diff
@@ -81,9 +69,32 @@ src_install() {
...
+	
+	# logrotate
+	insinto /etc/logrotate.d
+	newins "${FILESDIR}/i2pd-2.5.0.logrotate" i2pd
 }
 
 pkg_setup() {
 	enewgroup "${I2PD_GROUP}"
-	enewuser "${I2PD_USER}" -1 -1 "/var/lib/run/${PN}" "${I2PD_GROUP}"
+	enewuser "${I2PD_USER}" -1 -1 /var/lib/run/i2pd "${I2PD_GROUP}"
```
Comment (tomboy-64) [inline net-misc/i2pd/i2pd-2.6.0-r3.ebuild]: I trust you double-checked those changes for completeness. Because I haven't.

**Comment by rindeal** [comment]
Also please fix `--retry` in initd, which must not contain spaces. More info on this at https://github.com/OpenRC/openrc/issues/76.

**Comment by khumarahn** [comment]
Hi, thanks for the comments! I used `DOCS`, explicit octal mask `660`, removed spaces in `--retry`.

**Comment by khumarahn** [comment]
I added one more commit, increasing `ulimit -n` to 4096 for both openrc and systemd services. With default option of 1024 in floodfill regime the daemon becomes unresponsive, and control web page `http://localhost:7070` does not open. I updated the first post.

**Comment by blueness** [comment]
there have been a number of rapid changes to this package, some of them contradictory.  since i don't use it, and only proxy, i am not able to judge if they are improvements or not.  i'm going to slow down on these changes until the the changes settle down.

**Comment by tomboy-64** [comment]
Agreed.

@khumarahn please keep in mind that, apart from the time you invest to improve this ebuild, it takes quite some time to review it as well.

What I would suggest here is to save those changes in an additional branch; then upon the next version bump make a bump as simple as possible; then introducing these changes as an addition. All in one PR. First commit the bump; rest your changes.

**Comment by khumarahn** [comment]
@tomboy-64, @blueness, thank you for spending time on this. I learn from you. It is certainly reasonable to wait for changes to stabilize.

I would not recommend to postpone the current PR until the next release. This is because commits 0337b1c, 55a3684 and 6dda70b fix issues that may create problems for people using i2pd (like it happened to me). It would not be good to keep an ebuild for version 2.6.0 with this unsolved. And it wouldn't be reasonable to remove 2.6.0 too early, because we don't know if the next release will be stable.

Except 0337b1c, 55a3684 and 6dda70b, all other commits in this PR are cosmetic. I tried to make the ebuild cleaner so it is easier to maintain in the future.

Currently I can think only of one (future) change: in release 2.7.0 there will be an updated man page, and we may want to use it.

**Comment by tomboy-64** [comment]
fair enough.
i will have a closer look at them again tonight or tomorrow morning.

**Comment by khumarahn** [comment]
ping?

**Comment by tomboy-64** [comment]
Wow. Completely forgot adding that to my to-do list. Sorry about that. Getting on it.

**Comment by khumarahn** [comment]
ping?

**Comment by blueness** [comment]
@tomboy-64 are you finished reviewing this?

**Comment by tomboy-64** [comment]
I've been running I2Pd on 2 machines for the past week now, on arm32 as well as x86. I had problems with highly increased "crashiness" - after up to two hours, i2pd would just segfault.

On Wednesday I reverted the patch, re-run the original. Crashing was gone on both machines. Reinstalled the patches the same night, crashing came back on the arm machine. x86 ran fine.

I tried to properly debug and bisect it yesterday, but now (the third time not applying and then applying it) it's back in one of my x86 chroot installs (i made ~ half a dozen installs in x86 and amd64 chroots).

I'll give it another once-over. Since I can't reliably reproduce, please merge as a revbump and keep the old one then.

**Comment by blueness** [comment]
@khumarahn  can you redo this as a rev bump leaving intact all teh current config and init files.  thanks

**Comment by khumarahn** [comment]
Crashing on arm32 and x86 looks insteresting. I do not have access to x86. I ran it on amd64 for not very long, had no issues. On arm32 I ran it for longer:
```
Uptime: 495:16:43
Status: OK
Tunnel creation success rate: 33%
Received: 237.14 GiB (102.67 KiB/s)
Sent: 343.91 GiB (117.08 KiB/s)
```

ok, I will soon redo keeping the current revision intact

**Comment by blueness** [comment]
@khumarahn while i hate doing it, we can drop keywords for arches where it doesn't work.  so let's begin by introducing this as a rev bump until we figure out what's going on, if ever.  the problem is we have too small a user base to get a good feel for what's up

**Comment by tomboy-64** [comment]
I'm not sure it's serious enough to drop arches. As I said, I didn't manage to reliably replicate. I can't say it's a regression, either. Whether upstream- or Gentoo-wise or even depending on the machine it runs on.

I'm done with testing the proposed changes and all seems in order. I'll continue tracking those crashes.

One more question: what prompted you to do 55a3684? Afaict, both variants are valid.
Apart from this, these changes are a go.

**Comment by khumarahn** [comment]
About 55a3684: this was an advice from @rindeal above in this PR, see https://github.com/OpenRC/openrc/issues/76 It seems `--retry` option of `start-stop-daemon` is a bit underdocumented, and there should not be any spaces. It may work with spaces, but that may not be reliable.

About arches: I do testing on amd64 and arm32. I did not have any major problems.

I rewrote this PR, keeping all previous files intact. To make it easier to track, I am about to create a new PR, which will replace this one.

**Comment by khumarahn** [comment]
#1433 is a replacement, closing this

**Comment by rindeal** [comment]
as for https://github.com/gentoo/gentoo/commit/55a36845d37c7dfb2fe0c2d6755a4b024e98bec1, the relevant source code starts here: https://github.com/OpenRC/openrc/blob/0.20.5/src/rc/start-stop-daemon.c#L315. The code is messy, but if you read further, you will see, that the only reason, why the spaces do not issue errors, is that the parsing code simply ignores everything, after it parses a number or signal, up-to the next slash (https://github.com/OpenRC/openrc/blob/0.20.5/src/rc/start-stop-daemon.c#L262, https://github.com/OpenRC/openrc/blob/0.20.5/src/rc/start-stop-daemon.c#L240).

So the original `SIGTERM/20 SIGKILL/20` would silently issue `kill TERM`, `sleep 20`, `sleep 20`.

And there are more hidden things going on with OpenRC, so you have to be especially careful when writing OpenRC init scripts.

**Comment by khumarahn** [comment]
ok, interesting; so this is also a potential bugfix

**Pull request closed without merging**